boards/trellis_m4/src/orientation.rs

```diff
@@ -0,0 +1,187 @@
+//! Orientation tracking support for the NeoTrellis M4
+//!
+//! This module provides tablet-like orientation tracking. It's designed to
+//! work when the device is being handheld, in a "portrait" orientation,
+//! and attempts to detect whether the device's plugs (USB, audio, etc)
+//! are pointed up or down.
+
+pub use adxl343::Ax3;
+
+use crate::Adxl343;
+use hal::sercom::I2CError;
+
+/// Number of samples used to guess the device's orientation
+const NUM_SAMPLES: usize = 32;
+
+/// Orientation of the NeoTrellis M4, as detected by the onboard ADXL343
+/// accelerometer.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Orientation {
+    /// Device's plugs (USB, audio, etc) are pointed up
+    Up,
+
+    /// Device's plugs (USB, audio, etc) are pointed down
+    Down,
+
+    /// Unable to determine device's orientation from the current data
+    Unknown
+}
+
+impl Orientation {
+    /// Attempt to guess the orientation from the accelerometer data
```

Comment:
It really feels like there's a fairly trivial formula the implementation below is an extremely coarse grained approximation of. I will see if I can figure it out, but it seems like it should be possible to compute a single number which represents the devices up/down orientation  using the accelerometer data as inputs, and then it can be sliced into the appropriate `Orientation` based on its value.

Reply:
I will say I briefly looked at papers regarding this particular problem. There are a lot of them! Didn't find any specific magic formula, but it looks like I can probably convert the logic table below into corresponding arithmetic operations.